### PR TITLE
Icon button

### LINKED ISSUE #448: Text input v1.0.0
## Make text input component ready for publishing of v1.0.0

The following should be ready in the style (`jkl-text-input`) and react (`jkl-text-input-react`) packages before we can publish v1.0.0

### Style package should…
- [x] …have a `README.md` file containing _something_ that is **not** "TODO: …", preferrably simple usage instructions
- [x] …use only variables from `jkl-core` to define spacing, colors and font size/line height
- [x] …cover all variants found in the design sketches

### React package should…
- [x] …have a `README.md` file containing link to documentation on the portal
- [ ] …properly document all types/props in use and when/how to use them
- [x] …implement all variants present in style package
- [x] …implement all expected functionality (see below)
- [x] …generate WCAG-compliant and accessible HTML

### Expected functionality
- [x] Can have a presetvalue
- [x] Accepts help- and error messages
- [x] Accepts a placeholder
- [x] Accepts an `onChange` function
- [x] Accepts custom `className`s (e.g. for spacing)
- [x] Can be forced to be small even on large screens

### PR DESCRIPTION
**Pull request by frisol97**
## 📥 Proposed changes

There is a need for text-field with icons, so this is a suggestion of how it could be implemented. An action-icon acts like a button and should be used when it is necessary to make the use understand that this input-field (for example) is a search-field. 

This pull request adds action-icon and action-text-field components.

## ☑️ Submission checklist

-   [x] I have read the [CONTRIBUTING](https://github.com/fremtind/jokul/blob/master/CONTRIBUTING.md) doc
-   [x] `yarn build` works locally with my changes
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] I have added necessary documentation (if appropriate)

## 💬 Further comments

A better solution is needed.
1) We need a proposal from designer on how these icons should look.
2) The icons should not be a component themselves. Could we move it to the action-text-field or core?

closes #448

### CODE REVIEW

**Comment by piofinn** [comment]
@all-contributors, please add @vegardsoerlie for code!

File: package.json [hunk 1]
```diff
@@ -60,7 +60,7 @@
         "gulp-sass": "^4.0.2",
         "husky": "^3.0.0",
         "jest": "^24.8.0",
-        "lerna": "^3.16.4",
+        "lerna": "^3.17.0",
```
Comment (lfbergee) [inline package.json]: Please do upgrades like this in separate pr

File: packages/action-icon-react/src/ActionIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import React from "react";
+import { IconDelete } from "./Icons/IconDelete";
+import { IconEdit } from "./Icons/IconEdit";
+import { IconSearch } from "./Icons/IconSearch";
+import { IconClear } from "./Icons/IconClear";
+import { IconVariant } from "./index";
+
+interface Props {
+    name: IconVariant;
```
[the remark this answers is not in the document]
Comment (lfbergee) [reply packages/action-icon-react/src/ActionIcon.tsx]: If we want to contain this to just a few spesific actions, how about using the factory pattern instead? That makes it a bit more liberate to add another.

So <DeleteIconButton /> and <SearchIconButton />

File: packages/action-icon-react/src/ActionIcon.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+function ActionType(name: IconVariant) {
+    switch (name) {
+        case "clear":
+            return <IconClear />;
+            break;
```
Comment (lfbergee) [inline packages/action-icon-react/src/ActionIcon.tsx]: No need to break after return.

File: packages/action-icon-react/src/ActionIcon.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import { IconEdit } from "./Icons/IconEdit";
+import { IconSearch } from "./Icons/IconSearch";
+import { IconClear } from "./Icons/IconClear";
+import { IconVariant } from "./index";
+
+interface Props {
+    name: IconVariant;
+}
+
+function ActionType(name: IconVariant) {
```
[the remark this answers is not in the document]
Comment (lfbergee) [reply packages/action-icon-react/src/ActionIcon.tsx]: Well yeah, except functional react components, they must be capitalized. Normal functions do not :)

File: packages/action-icon-react/src/ActionIcon.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        case "edit":
+            return <IconEdit />;
+            break;
+        default:
+            return null;
+    }
+}
+
+export const ActionIcon = ({ name }: Props) => {
+    const componentClassName = "jkl-action-icon".concat(` jkl-action-icon--${name}`);
```
Comment (lfbergee) [inline packages/action-icon-react/src/ActionIcon.tsx]: Not really nessesary with the concatination here. It’s nice when we have multiple branches in one component, but this can just be a literal template

File: packages/action-icon-react/src/Icons/IconClear.tsx [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import React from "react";
+
+export const IconClear = () => (
+    <svg width="24" height="24" xmlns="http://www.w3.org/2000/svg" fillRule="evenodd" clipRule="evenodd">
```
Comment (lfbergee) [inline packages/action-icon-react/src/Icons/IconClear.tsx]: Maybe my pet peeve, but i think svgs should have a view box, no height, and take width as a argument with 24 as default argument

Comment (lfbergee) [reply packages/action-icon-react/src/Icons/IconClear.tsx]: The reasoning is that we can set the size of the icon dynamically to be any size we would like, just by changing a prop

File: packages/action-icon-react/src/Icons/IconSearch.tsx [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+
+export const IconSearch = () => (
+    <svg
+        style={{ color: "currentColor" }}
```
Comment (lfbergee) [inline packages/action-icon-react/src/Icons/IconSearch.tsx]: Inlinestyle?

File: packages/action-text-field-react/package.json [hunk 8]
```diff
@@ -0,0 +1,59 @@
+{
+    "name": "@fremtind/jkl-action-text-field-react",
+    "version": "0.0.0",
+    "publishConfig": {
+        "access": "public"
+    },
+    "description": "Jøkul react icon component",
```
Comment (lfbergee) [inline packages/action-text-field-react/package.json]: Copypasta?

File: packages/action-text-field-react/src/ActionTextField.tsx [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import React, { MouseEventHandler } from "react";
+import { TextField } from "@fremtind/jkl-text-input-react";
+import { ActionIcon, IconVariant } from "@fremtind/jkl-action-icon-react";
+import { Props as TextFieldProps } from "@fremtind/jkl-text-input-react/build/TextField";
```
Comment (lfbergee) [inline packages/action-text-field-react/src/ActionTextField.tsx]: From /build ?

File: packages/action-text-field/action-text-field.scss [hunk 10]
```diff
@@ -0,0 +1,27 @@
+@import "~@fremtind/jkl-core/variables/_all.scss";
+
+.jkl-text-field {
+    &__input {
+        padding-right: 3rem;
```
Comment (lfbergee) [inline packages/action-text-field/action-text-field.scss]: Might be a good idea to put these into a variable, makes it look a bit more deliberate

File: packages/action-icon-react/src/Icons/IconSearch.tsx (hunk 7 above)
Comment (piofinn) [reply packages/action-icon-react/src/Icons/IconSearch.tsx]: Yeah, that's my bad suggestion! 😅This can and should be set another way, maybe even in the component using the icon.

File: packages/action-icon-react/src/ActionIcon.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (piofinn) [reply packages/action-icon-react/src/ActionIcon.tsx]: Looks like your comment got a bit amputated there 😅I'm guessing the `Button` component is a good example of what you're suggesting? Seems like a good solution to consider!

[the remark this answers is not in the document]
Comment (lfbergee) [reply packages/action-icon-react/src/ActionIcon.tsx]: Yeah, on my iPad, can’t find backtics.... was DeleteIconButton and ClearIconButton I.e.

File: packages/action-icon-react/src/Icons/IconSearch.tsx (hunk 7 above)
Comment (lfbergee) [reply packages/action-icon-react/src/Icons/IconSearch.tsx]: But why only on the search icon, what is special about that? I do like inline styles every now and then.. but we might be able to solve this more elegantly.

Comment (vegardsoerlie) [reply packages/action-icon-react/src/Icons/IconSearch.tsx]: That one is on us, it was supposed to be implemented to all. Going to look into alternative solutions today amongst other things commented on in this PR.

File: package.json (hunk 1 above)
Comment (lfbergee) [reply package.json]: The reasoning for this is, unless the upgrade is nessesary to solve the task, this introduces extra risk in your PR. And can delay mergin the PR to deal with the extra risk. Handeling updates in the own PR makes it easier to revert if it fails in production. 

In Jøkul we have Greenkeeper, that might not work as it should at the moment, to upgrade dependecies and make PRs or issues if it fails or passes tests. Its a bit cleaner to let this be its own thing, istead of introducing it in a feature pr(unless you need some new function from a dependency)

File: packages/action-icon-react/src/ActionIcon.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (vegardsoerlie) [reply packages/action-icon-react/src/ActionIcon.tsx]: If we're implementing the factory pattern the complexity in ActionTextField increases. One solution could be to incorporate ActionIcon into ActionTextField so its not a stand-alone package.

[the remark this answers is not in the document]
Comment (lfbergee) [reply packages/action-icon-react/src/ActionIcon.tsx]: If ActionIcon only is used by ActionTextField and shouldn't be used by anyone else, it should be in the ActionTextField package. Not sure about the naming, I don't understand from the name what the difference between TextField and ActionTextField is.

I don't think we need to be scared about complexity, as long as we're aware of it and its there for a reason. In this case, we want to be really careful about adding more icons, so with a factory implementation it requires a bit more care to implement a new icon. So complexity becomes a feature.

File: packages/action-icon-react/src/Icons/IconSearch.tsx (hunk 7 above)
Comment (lfbergee) [reply packages/action-icon-react/src/Icons/IconSearch.tsx]: `fill=currentColor` i.e, or a scss file.

File: packages/action-text-field-react/src/ActionTextField.tsx [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    onClick: MouseEventHandler<HTMLButtonElement>;
+}
+
+export const ActionTextField = ({ iconType, onClick, className, ...rest }: Props) => {
+    return (
+        <div className="jkl-icon-text-field">
+            <button className="jkl-icon-text-field__icon" onClick={onClick} aria-label={iconType} title={iconType}>
+                <ActionIcon actionType={iconType} />
+            </button>
+            <TextField className={`jkl-icon-text-field__field ${className}`} {...rest} />
```
Comment (piofinn) [inline packages/action-text-field-react/src/ActionTextField.tsx]: `jkl-icon-text-field` should be changed to `jkl-action-text-field`

File: packages/action-text-field/action-text-field.scss [hunk 12]
```diff
@@ -0,0 +1,21 @@
+@import "~@fremtind/jkl-core/variables/_all.scss";
+
+.jkl-icon-text-field {
+    display: block;
+    position: relative;
+
+    &__field {
+        .jkl-text-field__input {
+            padding-right: 2.5rem;
```
Comment (piofinn) [inline packages/action-text-field/action-text-field.scss]: I believe `$component-spacing--xxl` can be used here for the same value :) Alternatively (and preferrably?) you should define the size of the icon somewhere with a variable, and define this padding using that variable and a spacing-unit. E.g.
```scss
$icon-size: rem(20px); // or something else
// ...
.jkl-text-field__input {
    padding-right: $icon-size + $component-spacing--small;
}
```

Makes it easoer to deduce the reason for the value.

File: packages/action-icon-react/src/ActionIcon.tsx [hunk 13]
```diff
@@ -0,0 +1,31 @@
+import React from "react";
```
Comment (lfbergee) [inline packages/action-icon-react/src/ActionIcon.tsx]: Hvorfor trenger vi egentlig denne fila? Den legger egentlig ikke til noe, annet enn at den samler ikonene i en React komponent. Kan vi ikke bare eksportere ikonene direkte?

File: packages/action-icon-react/src/Icons/IconClear.tsx [hunk 14]
```diff
@@ -0,0 +1,7 @@
+import React from "react";
+
+export const IconClear = () => (
```
Comment (lfbergee) [inline packages/action-icon-react/src/Icons/IconClear.tsx]: Ser gjerne at vi tar `width` som argument

File: packages/action-text-field-react/src/ActionTextField.tsx [hunk 15]
```diff
@@ -0,0 +1,20 @@
+import React, { MouseEventHandler } from "react";
+import { TextField } from "@fremtind/jkl-text-input-react";
+import { ActionIcon, IconVariant } from "@fremtind/jkl-action-icon-react";
+import { Props as TextFieldProps } from "@fremtind/jkl-text-input-react/build/TextField";
+
+interface Props extends TextFieldProps {
+    iconType: IconVariant;
```
Comment (lfbergee) [inline packages/action-text-field-react/src/ActionTextField.tsx]: Hvorfor ikke bare `icon`?

File: packages/action-icon-react/src/ActionIcon.tsx [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+import { IconDelete } from "./Icons/IconDelete";
+import { IconEdit } from "./Icons/IconEdit";
+import { IconSearch } from "./Icons/IconSearch";
+import { IconClear } from "./Icons/IconClear";
+import { IconVariant } from "./index";
+
+interface Props {
+    actionType: IconVariant;
```
Comment (lfbergee) [inline packages/action-icon-react/src/ActionIcon.tsx]: Hvorfor ikke bare `action`?

File: yarn.lock [hunk 17]
```diff
@@ -1222,13 +1222,13 @@
     "@types/istanbul-reports" "^1.1.1"
     "@types/yargs" "^13.0.0"
 
-"@lerna/add@3.16.5":
-  version "3.16.5"
-  resolved "https://registry.yarnpkg.com/@lerna/add/-/add-3.16.5.tgz#55f0aa4096d7c91c883ada073b4fa09e985fa9f8"
-  integrity sha512-vBG0G/w3yfbyaXSiOqRl86mRphBPPBinlPBqcFjxkuwvj58g/7gBBmBrji7WFO05eYvXVR0QZZik7NJ6qZSWWw==
+"@lerna/add@3.17.0":
```
Comment (lfbergee) [inline yarn.lock]: Dere må reverte yarn.lock fila også. Nå er den i usync siden package.json ble reversert.

File: packages/action-icon-react/src/Icons/IconClear.tsx (hunk 14 above)
Comment (piofinn) [reply packages/action-icon-react/src/Icons/IconClear.tsx]: Gir ikke det først mening hvis vi har noen definerte ikonstørrelser? Så fremt vi ikke har det er det vel bedre å la komponenten som bruker ikonet definere størrelsen via css..?

Comment (lfbergee) [reply packages/action-icon-react/src/Icons/IconClear.tsx]: Kan gjøres i css, jeg syns det er elegant å gjøre med props når man allerede har den i React.

```
const Svg = ({width=´100%´}) => <svg {width} ...
```

File: packages/action-text-field-react/src/ActionTextField.tsx (hunk 9 above)
Comment (vegardsoerlie) [reply packages/action-text-field-react/src/ActionTextField.tsx]: Det er midlertidig og alternativet blir å flytte nødvendige props ut i index-fila, men siden planen er å flytte komponentene over til text-input uansett tenker jeg det vi lar det vente til den endringen

File: packages/text-input-react/README.md [hunk 18]
```diff
@@ -2,9 +2,49 @@
...
+
+Du må spesifisere en `label` for samtlige komponenter omtalt under "Generell Bruk".
+
+Størrelsen på komponentene kan settes ved prop-en `variant`, hvor størrelsene er definert ved `"small" | "medium" | "large"`. Hvis komponentene skal brukes på mobile flater eller av andre årsaker trenger å være fremvist i sin kompakte variant må man ta i bruk `forceCompact`.
+
+Det er to måter å avgrense den fysiske størrelsen til inndata-feltet, enten ved å bruke `width` eller `maxLength`. `maxLength` avgrenser også hvor mye mange tegn som kan tastes inn. Hvis `maxLength` er definert og mindre enn `15` vil `width` bli satt til `${maxLength + 3}ch`.
+
+Komponentene tar imot hjelpe- og feilmeldinger gjennom `helpLabel` og `errorLabel`. Dersom `errorLabel` finnes vises _kun_ denne. Disse vil _ikke_ vises for komponenten `InlineTextField` og må fremvises på andre måter. Siden komponenten ikke har feil- og hjelpetekst brukes prop-en `invalid` for å sette feiltilstand.
+
+Handlingen/Ikonet som skal brukes på et `ActionTextField` gis gjennom prop-en `action` som igjen er et objekt bestående av `{{ icon, label, onClick }}`.
```
Comment (piofinn) [inline packages/text-input-react/README.md]: Her ville jeg kanskje definert objektet helt ut:
```ts
type action = {
  icon: IconVariant; // "clear", "delete", "search" eller "edit"
  label: string;
  onClick: (event: MouseEvent) => void;
}

File: packages/text-input-react/README.md [hunk 19]
```diff
@@ -2,9 +2,49 @@
...
+```js
+import { TextField, TextArea, ActionTextField, InlineTextField, BaseInputField } from "@fremtind/jkl-text-input-react";
+import "@fremtind/jkl-text-input/text-input.min.css";
+```
+
+### Generell Bruk
+
+Du må spesifisere en `label` for samtlige komponenter omtalt under "Generell Bruk".
+
+Størrelsen på komponentene kan settes ved prop-en `variant`, hvor størrelsene er definert ved `"small" | "medium" | "large"`. Hvis komponentene skal brukes på mobile flater eller av andre årsaker trenger å være fremvist i sin kompakte variant må man ta i bruk `forceCompact`.
```
Comment (piofinn) [inline packages/text-input-react/README.md]: Det er størrelsen på etiketten (label), ikke komponenten, som blir endret med `variant`. (kan skape forvirring ifm. `forceCompact`)

File: packages/text-input-react/README.md [hunk 20]
```diff
@@ -2,9 +2,49 @@
...
+2. Importer _både_ React-pakken (velg komponenten(e) du trenger) og stil-pakken i prosjektet ditt:
+
+```js
+import { TextField, TextArea, ActionTextField, InlineTextField, BaseInputField } from "@fremtind/jkl-text-input-react";
+import "@fremtind/jkl-text-input/text-input.min.css";
+```
+
+### Generell Bruk
+
+Du må spesifisere en `label` for samtlige komponenter omtalt under "Generell Bruk".
```
Comment (piofinn) [inline packages/text-input-react/README.md]: Kanskje omformulere denne litt siden variantene ikke listes opp lenger. Og kanskje spesifisere at `label` ikke vises for `InlineTextField`, men fortsatt brukes for skjermlesere? (bare så folk ikke er forvirra over at den er påkrevd, men ikke synes)

File: packages/text-input-react/src/ActionIcon.tsx [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+
+export const ActionIcon = ({ action, className }: Props) => {
+    return (
+        <span data-testid="jkl-action-icon" className={className}>
+            {getIcon(action)}
+        </span>
+    );
+};
+
+export type IconVariant = "clear" | "delete" | "search" | "edit";
```
Comment (lfbergee) [inline packages/text-input-react/src/ActionIcon.tsx]: Search og edit trengs vel ikke lengre?

Comment (vegardsoerlie) [reply packages/text-input-react/src/ActionIcon.tsx]: Heh nei, de lurte seg unna
